File: crates/openduo-agent/src/gitlab_provider.rs
use crate::provider::{
    ChatMessage, ChatRole, LlmProvider, ModelResponse, TokenStream, ToolDefinition,
};
use anyhow::{anyhow, Result};
use async_trait::async_trait;
use futures::StreamExt;
use openduo_core::{auth::AuthHeaders, config::Config};
use reqwest::Client;
use serde_json::{json, Value};
use tracing::{debug, error, instrument};

pub struct GitLabAiProvider {
    client: Client,
    gateway_url: String,
    pat: String,
}

impl GitLabAiProvider {
    pub fn new(config: &Config) -> Result<Self> {
        let client = Client::builder()
            .use_native_tls()
            .build()
            .map_err(|e| anyhow!("Failed to build reqwest client: {}", e))?;
        let gateway_url = format!(
            "{}/api/v4/chat/completions",
            config.gitlab_url.trim_end_matches('/')
        );
        Ok(Self {
            client,
            gateway_url,
            pat: config.pat.clone(),
        })
    }
}

#[async_trait]
impl LlmProvider for GitLabAiProvider {
    #[instrument(skip(self, messages, _tools))]
    async fn chat_stream(
        &self,
        messages: Vec<ChatMessage>,
        _tools: Vec<ToolDefinition>,
    ) -> Result<TokenStream> {
        // GitLab Duo Chat API uses Bearer auth, not PRIVATE-TOKEN
        let headers = AuthHeaders::new(&self.pat).to_bearer_header_map()?;

        // Extract the last user message as the content to send
        let content = messages
            .iter()
            .rev()
            .find(|m| matches!(m.role, ChatRole::User))
            .map(|m| m.content.clone())
            .unwrap_or_default();

        // GitLab Duo Chat API request format
        let body = json!({ "content": content });
        debug!(
            "Sending to GitLab Duo Chat: {} (content length: {})",
            self.gateway_url,
            content.len()
        );

        let raw_resp = self
            .client
            .post(&self.gateway_url)
            .headers(headers)
            .json(&body)
            .send()
            .await
            .map_err(|e| {
                error!(
                    "Failed to connect to GitLab Duo Chat at {}: {}",
                    self.gateway_url, e
                );
                anyhow!("Failed to connect to GitLab Duo Chat: {}", e)
            })?;

        let status = raw_resp.status();
        if !status.is_success() {
            let body_text = raw_resp.text().await.unwrap_or_default();
            error!("GitLab Duo Chat returned HTTP {}: {}", status, body_text);
            return Err(anyhow!(
                "GitLab Duo Chat returned HTTP {} — {}",
                status,
                body_text
            ));
        }

        let content_type = raw_resp
            .headers()
            .get(reqwest::header::CONTENT_TYPE)
            .and_then(|v| v.to_str().ok())
            .unwrap_or("")
            .to_string();
        debug!("Response content-type: {}", content_type);

        if content_type.contains("text/event-stream") {
            // SSE streaming response — parse data: lines
            let stream = raw_resp.bytes_stream().flat_map(move |chunk| {
                let events: Vec<Result<ModelResponse>> = match chunk {
                    Ok(bytes) => {
                        let text = String::from_utf8_lossy(&bytes).to_string();
                        parse_sse_chunk(&text)
                    }
                    Err(e) => vec![Err(anyhow!(e))],
                };
                futures::stream::iter(events)
            });
            Ok(Box::pin(stream))
        } else {
            // Non-streaming response — read entire body as text
            let body_text = raw_resp.text().await?;
            debug!("Duo Chat response length: {}", body_text.len());

            // Response may be a JSON string or plain text
            let response_text = if body_text.starts_with('"') {
                serde_json::from_str::<String>(&body_text).unwrap_or(body_text)
            } else {
                body_text
            };

            let events: Vec<Result<ModelResponse>> = vec![
                Ok(ModelResponse::Token(response_text)),
                Ok(ModelResponse::Done),
            ];
            Ok(Box::pin(futures::stream::iter(events)))
        }
    }
}

/// Parse SSE data: lines from a chunk, extracting text tokens.
fn parse_sse_chunk(text: &str) -> Vec<Result<ModelResponse>> {
    let mut events = Vec::new();
    let normalized = text.replace("\r\n", "\n");

    for line in normalized.lines() {
        if let Some(data) = line.strip_prefix("data: ") {
            let trimmed = data.trim();
            if trimmed == "[DONE]" || trimmed.is_empty() {
                events.push(Ok(ModelResponse::Done));
                continue;
            }
            // Try parsing as JSON first (OpenAI-compatible format)
            if let Ok(val) = serde_json::from_str::<Value>(trimmed) {
                // OpenAI format: choices[0].delta.content
                if let Some(token) = val["choices"][0]["delta"]["content"].as_str() {
                    if !token.is_empty() {
                        events.push(Ok(ModelResponse::Token(token.to_string())));
                    }
                }
                // GitLab format: may have "content" or "response" at top level
                else if let Some(token) = val["content"].as_str() {
                    if !token.is_empty() {
                        events.push(Ok(ModelResponse::Token(token.to_string())));
                    }
                } else if let Some(token) = val["response"].as_str() {
                    if !token.is_empty() {
                        events.push(Ok(ModelResponse::Token(token.to_string())));
                    }
                }
                if val["choices"][0]["finish_reason"].as_str() == Some("stop") {
                    events.push(Ok(ModelResponse::Done));
                }
            } else {
                // Plain text data line — treat as a token
                events.push(Ok(ModelResponse::Token(trimmed.to_string())));
            }
        }
    }
    events
}

File: extension/server/gitlabProvider.ts
import { bearerHeaders } from './auth';
import { classifyError, formatCauseChain, type Config, type DebugFlags } from './config';
import type { LlmProvider } from './provider';
import type { ChatMessage, ModelResponse, ToolDefinition } from './types';
import { flattenMessages } from './prompt';

export class GitLabAiProvider implements LlmProvider {
  private readonly gatewayUrl: string;
  private readonly pat: string;
  private readonly debug: DebugFlags;
  private readonly hasProxy: boolean;

  constructor(config: Config) {
    this.gatewayUrl = `${config.gitlabUrl.replace(/\/+$/, '')}/api/v4/chat/completions`;
    this.pat = config.pat;
    this.debug = config.debug;
    this.hasProxy = !!config.proxyUrl;
  }

  private debugLog(context: string, err: unknown): void {
    const categories = classifyError(err, this.hasProxy);
    const msg = err instanceof Error ? err.message : String(err);
    const code = (err as any)?.code ?? '';
    const stack = err instanceof Error ? err.stack ?? '' : '';
    const causeChain = formatCauseChain(err);

    for (const cat of categories) {
      if (this.debug[cat]) {
        console.log(`[${cat}] [rest] ${context}: ${msg}${code ? ` (code=${code})` : ''}`);
        if (causeChain) console.log(`[${cat}] [rest] cause chain:\n${causeChain}`);
        if (stack) console.log(`[${cat}] [rest] stack: ${stack}`);
      }
    }
    if (categories.length === 0 && this.debug.serverErrors) {
      console.log(`[serverErrors] [rest] ${context}: ${msg}`);
      if (causeChain) console.log(`[serverErrors] [rest] cause chain:\n${causeChain}`);
    }
  }

  async *chatStream(
    messages: ChatMessage[],
    _tools: ToolDefinition[],
  ): AsyncIterable<ModelResponse> {
    // Flatten the full conversation (system prompt + history + tool results)
    // into a single content string so the LLM sees the complete context,
    // including tool definitions and prior tool results.
    const content = flattenMessages(messages);

    let resp: Response;
    try {
      resp = await fetch(this.gatewayUrl, {
        method: 'POST',
        headers: bearerHeaders(this.pat),
        body: JSON.stringify({ content }),
      });
    } catch (err) {
      this.debugLog(`POST ${this.gatewayUrl}`, err);
      throw err;
    }

    if (!resp.ok) {
      const body = await resp.text().catch(() => '');
      const error = new Error(`GitLab Duo Chat returned HTTP ${resp.status} — ${body}`);
      if (this.debug.apiErrors) {
        console.log(`[apiErrors] [rest] POST ${this.gatewayUrl}: HTTP ${resp.status}`);
        console.log(`[apiErrors] [rest] response body: ${body.slice(0, 2000)}`);
        console.log(`[apiErrors] [rest] response headers: ${JSON.stringify(Object.fromEntries(resp.headers.entries()))}`);
      }
      throw error;
    }

    const contentType = resp.headers.get('content-type') ?? '';

    if (contentType.includes('text/event-stream')) {
      // SSE streaming response
      if (!resp.body) {
        throw new Error('No response body for SSE stream');
      }
      const reader = resp.body.getReader();
      const decoder = new TextDecoder();

      let buffer = '';
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        buffer += decoder.decode(value, { stream: true });

        const lines = buffer.split('\n');
        // Keep the last potentially incomplete line in the buffer
        buffer = lines.pop() ?? '';

        for (const line of lines) {
          const trimmed = line.trim();
          if (!trimmed.startsWith('data: ')) continue;
          const data = trimmed.slice(6).trim();

          if (data === '[DONE]' || data === '') {
            yield { type: 'done' };
            return;
          }

          // Try parsing as JSON
          try {
            const val = JSON.parse(data);
            // OpenAI format: choices[0].delta.content
            const deltaContent = val?.choices?.[0]?.delta?.content;
            if (deltaContent) {
              yield { type: 'token', token: deltaContent };
            }
            // GitLab format: "content" or "response" at top level
            else if (val?.content) {
              yield { type: 'token', token: val.content };
            } else if (val?.response) {
              yield { type: 'token', token: val.response };
            }

            if (val?.choices?.[0]?.finish_reason === 'stop') {
              yield { type: 'done' };
              return;
            }
          } catch {
            // Plain text data line — treat as a token
            yield { type: 'token', token: data };
          }
        }
      }
    } else {
      // Non-streaming response — read entire body
      let bodyText = await resp.text();
      // Response may be a JSON string or plain text
      if (bodyText.startsWith('"')) {
        try {
          bodyText = JSON.parse(bodyText) as string;
        } catch {
          // keep as-is
        }
      }
      yield { type: 'token', token: bodyText };
      yield { type: 'done' };
    }
  }
}
